dali/operators/geometry/affine_transforms/rotate_transform.cc

```diff
@@ -0,0 +1,147 @@
+)code")
+  .AddArg(
+    "angle",
+    R"code(Angle, in degrees.
+
+For two-dimensional data, the rotation is counter-clockwise, assuming the top-left corner is
+at ``(0,0)``. For three-dimensional data, the ``angle`` is a positive rotation around the provided
+axis.)code",
+    DALI_FLOAT, true)
+  .AddOptionalArg<std::vector<float>>(
+    "axis",
+    R"code(Applies **only** to three-dimension and is the axis around which to rotate the image.
+
+The vector does not need to be normalized, but it must have a non-zero length.
+
+Reversing the vector is equivalent to changing the sign of ``angle``.)code",
+    nullptr, true)
+  .AddOptionalArg<std::vector<float>>(
+    "center",
+    R"code(The center of the rotation operation.
+
+If provided, the number of elements should match the number of dimensions.)code",
+    nullptr, true)
+  .NumInput(0, 1)
+  .NumOutput(1)
+  .AddParent("TransformAttr");
+
+/**
+ * @brief Rotate transformation.
+ */
+class RotateTransformCPU
+    : public TransformBaseOp<CPUBackend, RotateTransformCPU> {
+ public:
+  using SupportedDims = dims<2, 3>;
+
+  explicit RotateTransformCPU(const OpSpec &spec) :
+      TransformBaseOp<CPUBackend, RotateTransformCPU>(spec),
+      angle_("angle", spec),
+      axis_("axis", spec),
+      center_("center", spec) {
+    assert(angle_.IsDefined());
+  }
+
+  /**
+   * @brief 2D rotation
+   */
+  template <typename T>
+  void DefineTransforms(span<affine_mat_t<T, 3>> matrices) {
+    constexpr int ndim = 2;
+    assert(matrices.size() == static_cast<int>(angle_.size()));
+    for (int i = 0; i < matrices.size(); i++) {
+      auto &mat = matrices[i];
+      auto angle = angle_[i];
+      mat = rotation2D(deg2rad(angle));
+
+      if (center_.IsDefined()) {
+        vec2 &center = *reinterpret_cast<vec2*>(center_[i].data());
+        mat.set_col(ndim, cat(sub<ndim, ndim>(mat) * -center + center, 1.0f));
+      }
+    }
+  }
+
+  /**
+   * @brief 3D rotation
+   */
+  template <typename T>
+  void DefineTransforms(span<affine_mat_t<T, 4>> matrices) {
+    constexpr int ndim = 3;
+    assert(matrices.size() == static_cast<int>(angle_.size()));
+    assert(matrices.size() == static_cast<int>(axis_.size()));
+    for (int i = 0; i < matrices.size(); i++) {
+      auto &mat = matrices[i];
+      auto angle = angle_[i];
+      vec3 &axis = *reinterpret_cast<vec3*>(axis_[i].data());
+      mat = rotation3D(axis, deg2rad(angle));
+
+      if (center_.IsDefined()) {
+        vec3 &center = *reinterpret_cast<vec3*>(center_[i].data());
+        mat.set_col(ndim, cat(sub<ndim, ndim>(mat) * -center + center, 1.0f));
+      }
+    }
+  }
+
+  void ProcessArgs(const OpSpec &spec, const workspace_t<CPUBackend> &ws) {
+    int repeat = IsConstantTransform() ? 1 : nsamples_;
+    angle_.Read(spec, ws, repeat);
+    ndim_ = axis_.IsDefined() ? 3 : 2;
+    if (axis_.IsDefined()) {
+      axis_.Read(spec, ws, repeat);
```

Comment:
I wonder if Read can check if the argument is defined without asking to do that explicitly.
Also I don't know if we can generalize this enforce and make it a part of read itself.

Reply:
Read will do nothing if the argument is not defined. However, I want to check if the number of dimensions is correct. That's why I have this IsDefined() guard mainly